crates/pgt_workspace/src/workspace/server/parser.rs

```diff
@@ -0,0 +1,374 @@
+        content: &str,
+    ) -> Self::Output {
+        (id, range, content.to_string())
+    }
+}
+
+pub struct ExecuteStatementMapper;
+impl<'a> StatementMapper<'a> for ExecuteStatementMapper {
+    type Output = (
+        StatementId,
+        TextRange,
+        String,
+        Option<pgt_query_ext::NodeEnum>,
+    );
+
+    fn map(
+        &self,
+        parser: &'a Parser,
+        id: StatementId,
+        range: TextRange,
+        content: &str,
+    ) -> Self::Output {
+        let ast_result = parser.ast_db.load_parse(&id, content);
+        let ast_option = match &*ast_result {
+            Ok(node) => Some(node.clone()),
+            Err(_) => None,
+        };
+
+        (id, range, content.to_string(), ast_option)
+    }
+}
+
+pub struct AsyncDiagnosticsMapper;
+impl<'a> StatementMapper<'a> for AsyncDiagnosticsMapper {
+    type Output = (
+        StatementId,
+        TextRange,
+        String,
+        Option<pgt_query_ext::NodeEnum>,
+        Arc<tree_sitter::Tree>,
+    );
+
+    fn map(
+        &self,
+        parser: &'a Parser,
+        id: StatementId,
+        range: TextRange,
+        content: &str,
+    ) -> Self::Output {
+        let content_owned = content.to_string();
+        let ast_result = parser.ast_db.load_parse(&id, &content_owned);
+
+        let ast_option = match &*ast_result {
+            Ok(node) => Some(node.clone()),
+            Err(_) => None,
+        };
+
+        let cst_result = parser.cst_db.load_parse(&id, &content_owned);
+
+        (id, range, content_owned, ast_option, cst_result)
+    }
+}
+
+pub struct SyncDiagnosticsMapper;
+impl<'a> StatementMapper<'a> for SyncDiagnosticsMapper {
+    type Output = (
+        StatementId,
+        TextRange,
+        Option<pgt_query_ext::NodeEnum>,
+        Option<SyntaxDiagnostic>,
+    );
+
+    fn map(
+        &self,
+        parser: &'a Parser,
+        id: StatementId,
+        range: TextRange,
+        content: &str,
+    ) -> Self::Output {
+        let ast_result = parser.ast_db.load_parse(&id, content);
+
+        let (ast_option, diagnostics) = match &*ast_result {
+            Ok(node) => (Some(node.clone()), None),
+            Err(diag) => (None, Some(diag.clone())),
+        };
+
+        (id, range, ast_option, diagnostics)
+    }
+}
+
+pub struct GetCompletionsMapper;
+impl<'a> StatementMapper<'a> for GetCompletionsMapper {
+    type Output = (StatementId, TextRange, String, Arc<tree_sitter::Tree>);
+
+    fn map(
+        &self,
+        parser: &'a Parser,
+        id: StatementId,
+        range: TextRange,
+        content: &str,
+    ) -> Self::Output {
+        let cst_result = parser.cst_db.load_parse(&id, content);
+        (id, range, content.to_string(), cst_result)
+    }
+}
+
+pub struct DefaultFilter;
+impl<'a> StatementFilter<'a> for DefaultFilter {
+    fn apply(&self, _id: &StatementId, _range: &TextRange) -> bool {
+        true
+    }
+}
+
+pub struct CursorPositionFilter {
+    pos: TextSize,
+}
```

Comment:
love this!
I'm already seeing a `CursorPositionWithOffsetFilter` for the completions